File: ml-service/app/schemas/topic.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from app.schemas.document import CamelModel

GroundingLabel = Literal["supported", "unsupported", "partial"]


class ArxivCandidate(CamelModel):
    arxiv_id: str
    title: str
    authors: list[str]
    published: str
    abstract: str
    score: float  # similarity of the paper's title + abstract to the topic
    is_recent: bool = False  # published within the last two years


class BriefItem(CamelModel):
    id: str
    text: str
    quote: str
    page: int
    retrieval_score: float
    grounding_label: GroundingLabel | None
    grounding_score: float | None


class PaperBrief(CamelModel):
    problem: list[BriefItem] = []
    method: list[BriefItem] = []
    results: list[BriefItem] = []
    matters: list[BriefItem] = []


class TopicPaperOut(CamelModel):
    id: str
    arxiv_id: str
    title: str
    authors: list[str]
    published: str
    abstract: str
    status: Literal["pending", "ingesting", "analyzing", "done", "error"]
    error: str | None
    document_id: str | None
    brief_mode: str | None
    brief: PaperBrief


class TimelineEntry(CamelModel):
    paper_id: str
    label: str
    title: str
    year: str
    published: str


class RelationshipOut(CamelModel):
    from_paper_id: str
    to_paper_id: str
    kind: Literal["cites", "similar"]
    explanation: str
    page: int | None = None
    snippet: str | None = None
    shared_terms: list[str] = []
    similarity: float | None = None


class HowTheyFit(CamelModel):
    paper_ids: list[str]
    text: str


class ComparisonRow(CamelModel):
    paper_id: str
    label: str
    title: str
    year: str
    problem: str
    method: str
    results: str
    matters: str


class TopicSynthesis(CamelModel):
    mode: str  # "template" or "llm:<provider>"
    overview: str
    reading_order: list[str]
    timeline: list[TimelineEntry]
    relationships: list[RelationshipOut]
    how_they_fit: list[HowTheyFit]
    comparison: list[ComparisonRow]
    foundational: list[str]


class TopicOut(CamelModel):
    id: str
    query: str
    status: Literal["searching", "ingesting", "analyzing", "synthesizing", "done", "error"]
    error: str | None
    created_at: datetime
    papers: list[TopicPaperOut]
    synthesis: TopicSynthesis | None


class TopicSummary(CamelModel):
    id: str
    query: str
    status: str
    paper_count: int
    created_at: datetime


# ---- request bodies (snake_case, like the other routers' request models) ----


class PaperMeta(BaseModel):
    arxiv_id: str
    title: str
    authors: list[str] = []
    published: str = ""
    abstract: str = ""


class SearchTopicRequest(BaseModel):
    query: str
    limit: int = 5


class CreateTopicRequest(BaseModel):
    query: str
    max_papers: int = 5
    # Papers the user picked from a search preview; omitted = auto-pick the top matches.
    papers: list[PaperMeta] | None = None

File: app/src/lib/schemas/topic.ts
import { z } from "zod";
import { GroundingLabelSchema } from "./document";

/** Mirrors ml-service/app/schemas/topic.py - keep the two in sync by hand. */

export const ArxivCandidateSchema = z.object({
  arxivId: z.string(),
  title: z.string(),
  authors: z.array(z.string()),
  published: z.string(),
  abstract: z.string(),
  score: z.number(),
  isRecent: z.boolean(),
});
export type ArxivCandidate = z.infer<typeof ArxivCandidateSchema>;

export const BriefItemSchema = z.object({
  id: z.string(),
  text: z.string(),
  quote: z.string(),
  page: z.number().int(),
  retrievalScore: z.number(),
  groundingLabel: GroundingLabelSchema.nullable(),
  groundingScore: z.number().nullable(),
});
export type BriefItem = z.infer<typeof BriefItemSchema>;

export const PaperBriefSchema = z.object({
  problem: z.array(BriefItemSchema),
  method: z.array(BriefItemSchema),
  results: z.array(BriefItemSchema),
  matters: z.array(BriefItemSchema),
});
export type PaperBrief = z.infer<typeof PaperBriefSchema>;

export const PaperStatusSchema = z.enum(["pending", "ingesting", "analyzing", "done", "error"]);
export type PaperStatus = z.infer<typeof PaperStatusSchema>;

export const TopicPaperSchema = z.object({
  id: z.string(),
  arxivId: z.string(),
  title: z.string(),
  authors: z.array(z.string()),
  published: z.string(),
  abstract: z.string(),
  status: PaperStatusSchema,
  error: z.string().nullable(),
  documentId: z.string().nullable(),
  briefMode: z.string().nullable(),
  brief: PaperBriefSchema,
});
export type TopicPaper = z.infer<typeof TopicPaperSchema>;

export const RelationshipSchema = z.object({
  fromPaperId: z.string(),
  toPaperId: z.string(),
  kind: z.enum(["cites", "similar"]),
  explanation: z.string(),
  page: z.number().int().nullable(),
  snippet: z.string().nullable(),
  sharedTerms: z.array(z.string()),
  similarity: z.number().nullable(),
});
export type Relationship = z.infer<typeof RelationshipSchema>;

export const TopicSynthesisSchema = z.object({
  mode: z.string(),
  overview: z.string(),
  readingOrder: z.array(z.string()),
  timeline: z.array(
    z.object({
      paperId: z.string(),
      label: z.string(),
      title: z.string(),
      year: z.string(),
      published: z.string(),
    }),
  ),
  relationships: z.array(RelationshipSchema),
  howTheyFit: z.array(z.object({ paperIds: z.array(z.string()), text: z.string() })),
  comparison: z.array(
    z.object({
      paperId: z.string(),
      label: z.string(),
      title: z.string(),
      year: z.string(),
      problem: z.string(),
      method: z.string(),
      results: z.string(),
      matters: z.string(),
    }),
  ),
  foundational: z.array(z.string()),
});
export type TopicSynthesis = z.infer<typeof TopicSynthesisSchema>;

export const TopicStatusSchema = z.enum([
  "searching",
  "ingesting",
  "analyzing",
  "synthesizing",
  "done",
  "error",
]);
export type TopicStatus = z.infer<typeof TopicStatusSchema>;

export const TopicSchema = z.object({
  id: z.string(),
  query: z.string(),
  status: TopicStatusSchema,
  error: z.string().nullable(),
  createdAt: z.string(),
  papers: z.array(TopicPaperSchema),
  synthesis: TopicSynthesisSchema.nullable(),
});
export type Topic = z.infer<typeof TopicSchema>;

export const TopicSummarySchema = z.object({
  id: z.string(),
  query: z.string(),
  status: z.string(),
  paperCount: z.number().int(),
  createdAt: z.string(),
});
export type TopicSummary = z.infer<typeof TopicSummarySchema>;
